### PR TITLE
utility to send pinned xsk_map to cndp

### PR DESCRIPTION
**Pull request by maryamtahhan**
This commit introduces a utility that will allow a user to send
a pinned xsk_map to a CNDP application. It allows the user to
write their own BPF programs and load them with their preferred
BPF program loader. The only constraint is that they must pin
their XSK_MAP.

The program is currently limited to sending a single XSK_MAP FD
to CNDP.

### CODE REVIEW

File: usrtools/xskmap_load_send/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# xsk_map load and send utility
+
+This is a simply utility that loads a pinned xsk_map fd and sends it to
```
Comment (maryamtahhan) [inline usrtools/xskmap_load_send/README.md:3]:
```suggestion
This is a simple utility that loads a pinned xsk_map fd and sends it to
```

File: usrtools/xskmap_load_send/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+a CNDP application that's configured to work in unprivileged mode. It
+allows the end user to load their own custom BPF programs besides the
+default AF_XDP redirect program provided by the kernel.
+For e.g. https://github.com/maryamtahhan/xdp-progs/blob/main/xdp-filter-udp/xdp_prog_kern.c
+
+> **_NOTE_** The loading is left outside the scope of CNDP as there are
+many options available for use.
+
+Running the application will create a UDS named `app_socket.{proc_id}`
+in the path `/var/run/cndp/`. An example configuration for the `cndpfwd`
```
Comment (maryamtahhan) [inline usrtools/xskmap_load_send/README.md:13]:
```suggestion
in the path `/var/run/cndp/`. An example configuration for the `cndpfwd` application to be used alongside this utility
```

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    return 0;
+}
+
+static int
+send_map_fd(int sock, int fd)
+{
+    char cmsgbuf[CMSG_SPACE(sizeof(int))];
+    struct msghdr msg;
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:76]: Should we set cmsgbuf[] = {0}; and msg = {0}; to be safe here?

We could leak data in at least cmsgbuf[]

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+    cmsg->cmsg_level = SOL_SOCKET;
+    cmsg->cmsg_type  = SCM_RIGHTS;
+    cmsg->cmsg_len   = CMSG_LEN(sizeof(int));
+
+    *(int *)CMSG_DATA(cmsg) = fd;
+    int ret                 = sendmsg(sock, &msg, 0);
+
+    if (ret < 0) {
+        CNE_ERR("Failed to send xsk_map fd with error=%s\n", strerror(errno));
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:103]: You could use CNE_ERR_RET_VAL(-errno, "Failed to send ....); but only to return the it to one liner.

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+static int
+send_map_fd(int sock, int fd)
+{
+    char cmsgbuf[CMSG_SPACE(sizeof(int))];
+    struct msghdr msg;
+    struct iovec iov;
+    char value[UDS_MAX_CMD_LEN] = {0};
+
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:79]: Do we need to test sock and fd to be valid?

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        return -errno;
+    }
+
+    return ret;
+}
+
+static int
+connect_host(uds_client_t *c, const char *cmd __cne_unused, const char *params __cne_unused)
+{
+
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:113]: Do we need to check 'c' is valid or is that done in a previous call?

If we need to check it here then the other routines need to check.

Also remove blank line.

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    else
+        CNE_DEBUG("Sent %s msg\n", UDS_FIN_ACK_MSG);
+
+    return 0;
+}
+
+static int
+send_xskmap_fd(uds_client_t *c, const char *cmd __cne_unused, const char *params __cne_unused)
+{
+    send_map_fd(c->s, c->info->xsk_map_fd);
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:136]: Need to test the return value here.

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+        fwd->timer_quit = 1;
+        break;
+
+    case CNE_CALLED_EXIT:
+        if (val)
+            cne_printf_pos(99, 1, "\n>>> [cyan]Terminating with status [green]%d[]\n", val);
+
+        if (fwd) {
+            cne_printf(">>> [magenta]Closing\n");
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:164]: Need the tailing [] before \n to terminate the color. Then the line at  176 may not be needed.

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    int fd;
+
+    if (cne_init() || parse_args(argc, argv))
+        return -1;
+
+    cne_on_exit(__on_exit, (void *)&info, signals, cne_countof(signals));
+
+    fd = bpf_obj_get(info.map_path);
+    if (fd < 0) {
+        CNE_ERR_RET("Failed to open pinned xsk_map:%s err:%s\n", info.map_path, strerror(errno));
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:195]: remove {} you have been writing too much Go code :-)

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    CNE_DEBUG("xsk_map fd =%d\n", fd);
+
+    info.uds_info = uds_get_default(&info);
+    if (!info.uds_info) {
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.c:200]: Here too

File: usrtools/xskmap_load_send/xskmap_load_send.h [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+#define OPT_NO_COLOR     "no-color"
+#define OPT_NO_COLOR_NUM 256
+
+struct map_info {
+    char map_path[1024];
+    uds_info_t *uds_info;
+    volatile int timer_quit;
+};
+
+struct map_info info;
```
Comment (KeithWiles) [inline usrtools/xskmap_load_send/xskmap_load_send.h:29]: Need to add a newline after this line.

**Review by KeithWiles — CHANGES_REQUESTED** [review]

File: usrtools/xskmap_load_send/meson.build [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+if cne_verbose
+	message('Build xskmap_load_send')
+endif
+
+deps += [bpf_dep, uds]
+
+sources = files('xskmap_load_send.c')
+
+jsonc2json = executable('xskmap_load_and_send', sources, dependencies: deps, install: true)
```
Comment (jeffreybshaw) [inline usrtools/xskmap_load_send/meson.build:12]: typo: jsonc2json

File: usrtools/xskmap_load_send/xskmap_load_send.h (hunk 11 above)
Comment (jeffreybshaw) [inline usrtools/xskmap_load_send/xskmap_load_send.h:29]: shoudl add newline at the end of hte file

File: usrtools/xskmap_load_send/xskmap_load_send.c (hunk 7 above)
Comment (jeffreybshaw) [inline usrtools/xskmap_load_send/xskmap_load_send.c:136]: Do you want to return send_map_fd()?

File: usrtools/xskmap_load_send/xskmap_load_send.c [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+    int ret                 = sendmsg(sock, &msg, 0);
+
+    if (ret < 0)
+        CNE_ERR_RET_VAL(-errno, "Failed to send xsk_map fd with error=%s\n", strerror(errno));
+
+    return ret;
+}
+
+static int
+connect_host(uds_client_t *c, const char *cmd __cne_unused, const char *params __cne_unused)
```
Comment (maryamtahhan) [inline usrtools/xskmap_load_send/xskmap_load_send.c:109]: re checking uds_client_t *c -- I believe this is done through the client handler https://github.com/CloudNativeDataPlane/cndp/blob/main/lib/common/uds/uds.c#L475

File: usrtools/xskmap_load_send/xskmap_load_send.c (hunk 6 above)
Comment (maryamtahhan) [reply usrtools/xskmap_load_send/xskmap_load_send.c]: re checking uds_client_t *c -- I believe this is done through the client handler https://github.com/CloudNativeDataPlane/cndp/blob/main/lib/common/uds/uds.c#L475

File: usrtools/xskmap_load_send/xskmap_load_send.c (hunk 7 above)
Comment (maryamtahhan) [reply usrtools/xskmap_load_send/xskmap_load_send.c]: going to forward the return value through to the calling function through the client handler

**Review by KeithWiles — APPROVED** [review]
LGTM

**Pull request merged**